presto-iceberg/src/main/java/com/facebook/presto/iceberg/changelog/ChangelogPageSource.java

```diff
@@ -0,0 +1,160 @@
+    private final List<IcebergColumnHandle> desiredColumns;
+    private final List<IcebergColumnHandle> delegateColumns;
+    private final RowType innerRowType;
+
+    public ChangelogPageSource(ConnectorPageSource delegate, ChangelogSplitInfo changelogSplitInfo, ConnectorSplit split, List<IcebergColumnHandle> desiredColumns, List<IcebergColumnHandle> delegateColumns)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+        this.split = requireNonNull(split, "split is null");
+        this.changelogSplitInfo = requireNonNull(changelogSplitInfo, "changelogSplitInfo is null");
+        this.desiredColumns = requireNonNull(desiredColumns, "columns is null");
+        this.delegateColumns = requireNonNull(delegateColumns, "delegateColumns is null");
+        this.innerRowType = RowType.from(delegateColumns.stream().map(x -> new RowType.Field(Optional.of(x.getName()), x.getType())).collect(Collectors.toList()));
+//        this.primaryKeyIndex = IntStream.range(0, delegateColumns.size())
+//                .filter(i -> delegateColumns.get(i).getName().equals(changelogSplitInfo.getPrimaryKeyColumnName())).findFirst().getAsInt();
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return delegate.getCompletedBytes();
+    }
+
+    @Override
+    public long getCompletedPositions()
+    {
+        return delegate.getCompletedPositions();
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return delegate.getReadTimeNanos();
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return delegate.isFinished();
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        // In order to produce the correct page, we have to build the blocks that
+        // make up the schema of the changelog table.
+        // First, create blocks matching the selected columns in the page
+        // if required, extract the block pertaining to the primary key of the
+        // table then, combine all blocks into a single page.
+        Page delegatePage = delegate.getNextPage();
+        if (delegatePage == null) {
+            return null;
+        }
+
+        Block[] columns = new Block[desiredColumns.size()];
+        for (int i = 0; i < columns.length; i++) {
+            columns[i] = ChangelogSchemaColumns.valueOf(desiredColumns.get(i).getName().toUpperCase()).getBlock(this, delegatePage);
+        }
+
+        return new Page(delegatePage.getPositionCount(), columns);
+    }
+
+    @Override
+    public long getSystemMemoryUsage()
+    {
+        return delegate.getSystemMemoryUsage();
+    }
+
+    @Override
+    public void close()
+            throws IOException
+    {
+        delegate.close();
+    }
+
+    public enum ChangelogSchemaColumns
+    {
+        OPERATION((source, page) -> RunLengthEncodedBlock.create(VARCHAR, Slices.utf8Slice(source.changelogSplitInfo.getOperation().toString()), page.getPositionCount())),
```

Comment:
ChangelogOperation appears to have two flavors of update - UPDATE_BEFORE and UPDATE_AFTER. Can you describe the difference between them?

Reply:
[An example exists in the Iceberg documentation](https://iceberg.apache.org/docs/latest/spark-procedures/#prepost-update-images).

Essentially `UPDATE_BEFORE` and `UPDATE_AFTER` amount to is that they reflect the value of a row before an update and after an update within a particular ordinal. Whether updates are encoded like this depend on the table and Spark configuration used to perform the update.

As explained in the docs, if you had an update encoded as

| id | name | operation |
|-|-|-|
| 3 | Robert | DELETE |
| 3 | Dan | INSERT |

Spark could also encode the same information as:


| id | name | operation |
|-|-|-|
| 3 | Robert | UPDATE_BEFORE |
| 3 | Dan | UPDATE_AFTER |